File: 2021/java/Day0101.java
class Day0101 {
    public static void main(String[] args) {
        System.out.println("nth_even #1: " + nthEven(1));
        System.out.println("nth_even #2: " + nthEven(2));
        System.out.println("nth_even #3: " + nthEven(100));
        System.out.println("-------------");
        System.out.println("validateEmail #1: " + validateEmail("@gmail.com"));
        System.out.println("validateEmail #2: " + validateEmail("hello.gmail@com"));
        System.out.println("validateEmail #3: " + validateEmail("gmail"));
        System.out.println("validateEmail #4: " + validateEmail("hello@gmail"));
        System.out.println("validateEmail #5: " + validateEmail("hello@edabit.com"));
        System.out.println("-------------");
        System.out.println("atbash #1: " + atbash("apple"));
        System.out.println("atbash #2: " + atbash("Hello world!"));
        System.out.println("atbash #3: " + atbash("Christmas is the 25th of December"));
    }
    
    private static int nthEven(int n) {
        return 2 * n - 2;
    }
    
    private static boolean validateEmail(String email) {
        return email.matches("^.+@.+\\..+$");
    }
    
    private static String atbash(String s) {
        StringBuilder res = new StringBuilder();
        for (char c : s.toCharArray()) {
            if (Character.isLetter(c)) {
                if (Character.isUpperCase(c)) {
                    res.appendCodePoint('Z' - c + 'A');
                } else {
                    res.appendCodePoint('z' - c + 'a');
                }
            } else {
                res.append(c);
            }
        }
        return res.toString();
    }
}

File: 2021/kotlin/Day0101.kt
fun nthEven(a: Int) = a * 2 - 2
fun validateEmail(email: String) = email.matches("^.+@.+\\..+$".toRegex())

fun Char.toAtBash(): Char {
    var r = 'z' - this.toLowerCase().toInt() + 'a'.toInt()
    if (this.isUpperCase()) r = r.toUpperCase()
    return r
}

fun atBash(s: String): String {
    var result = ""
    for (c in s) {
        if (c.isLetter()) {
            result += c.toAtBash()
            continue
        }
        result += c
    }
    return result
}

fun main() {
    println(nthEven(1))
    println(nthEven(2))
    println(nthEven(100))
    println("-----")
    println(validateEmail("@gmail.com"))
    println(validateEmail("hello.gmail@com"))
    println(validateEmail("gmail"))
    println(validateEmail("hello@gmail"))
    println(validateEmail("hello@edabit.com"))
    println("-----")
    println(atBash("apple"))
    println(atBash("Hello world!"))
    println(atBash("Christmas is the 25th of December"))
}
